packages/next-on-pages/templates/cache/incrementalCache.ts

```diff
@@ -0,0 +1,226 @@
+ * source: https://github.com/vercel/next.js/blob/c4adae89b/packages/next/src/server/lib/incremental-cache/index.ts#L65
+ */
+export class IncrementalCache {
+	cacheHandler: CacheHandler;
+	requestHeaders: Record<string, undefined | string | string[]>;
+	revalidatedTags?: string[];
+
+	constructor({
+		curCacheHandler,
+		requestHeaders,
+		fetchCacheKeyPrefix,
+	}: {
+		allowedRevalidateHeaderKeys?: string[];
+		requestHeaders: IncrementalCache['requestHeaders'];
+		fetchCacheKeyPrefix?: string;
+		curCacheHandler: typeof CacheHandler;
+	}) {
+		this.requestHeaders = requestHeaders;
+
+		let revalidatedTags: string[] = [];
+
+		if (
+			typeof requestHeaders[NEXT_CACHE_REVALIDATED_TAGS_HEADER] === 'string'
+		) {
+			revalidatedTags =
+				requestHeaders[NEXT_CACHE_REVALIDATED_TAGS_HEADER].split(',');
+		}
+
+		this.cacheHandler = new curCacheHandler({
+			revalidatedTags,
+			_requestHeaders: requestHeaders,
+			fetchCacheKeyPrefix,
+		});
+	}
+
+	async revalidateTag(tag: string) {
+		return this.cacheHandler?.revalidateTag(tag);
+	}
+
+	async get(
+		cacheKey: string,
+		ctx: {
+			kindHint?: IncrementalCacheKindHint;
+			revalidate?: number | false;
+			fetchUrl?: string;
+			fetchIdx?: number;
+			tags?: string[];
+			softTags?: string[];
+		} = {},
+	): Promise<(IncrementalCacheEntry & { age?: number }) | null> {
+		let entry: (IncrementalCacheEntry & { age?: number }) | null = null;
+		let revalidate = ctx.revalidate;
+
+		const cacheData = await this.cacheHandler.get(cacheKey, ctx);
+
+		const age = cacheData
+			? (Date.now() - (cacheData.lastModified || 0)) / 1000
+			: undefined;
+
+		if (cacheData?.value?.kind === 'FETCH') {
+			const combinedTags = [...(ctx.tags || []), ...(ctx.softTags || [])];
+			// if a tag was revalidated we don't return stale data
+			if (
+				combinedTags.some(tag => {
+					return this.revalidatedTags?.includes(tag);
+				})
+			) {
+				return null;
+			}
+
+			revalidate = revalidate || cacheData.value.revalidate;
+			const isStale =
+				typeof revalidate === 'number' &&
+				typeof age === 'number' &&
+				age > revalidate;
+			const data = cacheData.value.data;
+
+			return {
+				isStale: isStale,
+				value: {
+					kind: 'FETCH',
+					data,
+					revalidate: revalidate,
+				},
+				age,
+				revalidateAfter:
+					typeof revalidate === 'number' && Date.now() + revalidate * 1000,
+			};
+		}
+
+		let isStale: boolean | -1 | undefined;
+		let revalidateAfter: false | number;
+
+		if (cacheData?.lastModified === -1) {
+			revalidateAfter = -1 * CACHE_ONE_YEAR;
+			isStale = -1;
+		} else {
+			revalidateAfter = 1 * 1000 + (cacheData?.lastModified || Date.now());
+			isStale = revalidateAfter < Date.now() ? true : undefined;
+		}
+
+		entry = {
+			isStale,
+			revalidateAfter,
+			value: null,
+		};
+
+		if (cacheData) {
+			entry.value = cacheData.value;
+			entry.age = age;
+		} else {
+			await this.set(cacheKey, entry.value, ctx);
+		}
```

Comment:
Note: I think this is incorrect

if `cacheData` is falsy we end up setting the `cacheKay` value to `null`  here, which generates the error mentioned in https://github.com/cloudflare/next-on-pages/pull/606#issuecomment-1904813309

Reply:
But the exact same happens in the Next.js code... 🤔 

https://github.com/vercel/next.js/blob/fc7a28a0dd946437cbd5144a57f5296e2659899f/packages/next/src/server/lib/incremental-cache/index.ts#L524-L530